File: governor_model/governor/governor.cpp
//
// Created by ocelot on 1/29/24.
//
#include <iostream>
#include <cstdlib>
#include <stdio.h>
#include <stdlib.h>
#include <sstream>
#include <fstream>
#include <array>

#include "GA.h"
#include "fitness.h"

using namespace std;

int LittleFrequencyTable[] = { 500000, 667000, 1000000, 1200000, 1398000, 1512000, 1608000, 1704000, 1800000 };
int BigFrequencyTable[]    = { 500000,  667000,  1000000, 1200000, 1398000, 1512000, 1608000,
                               1704000, 1800000, 1908000, 2016000, 2100000, 2208000 };

const float NUDGE = 1.3;
const float ANTI_NUDGE = 0.4;
const float ANTI_NUDGE_THRESH = 0.04;
const int REPETITION_LIMIT = 4;

void setup() {
  /* Export OpenCL library path */
  system("export LD_LIBRARY_PATH=/data/local/Working_dir");
  setenv("LD_LIBRARY_PATH", "/data/local/Working_dir", 1);

  /* Setup Performance Governor (CPU) */
  system("echo performance > /sys/devices/system/cpu/cpufreq/policy0/scaling_governor");
  system("echo performance > /sys/devices/system/cpu/cpufreq/policy2/scaling_governor");

  /* set fan speed to 100% */
  system("echo 1 > /sys/class/fan/enable");
  system("echo 0 > /sys/class/fan/mode");
  system("echo 4 > /sys/class/fan/level");
}

void scale_little(int freq) {
  string Command = "echo " + to_string(freq) + " > /sys/devices/system/cpu/cpufreq/policy0/scaling_max_freq";
  system(Command.c_str());
}

void scale_big(int freq) {
  string Command = "echo " + to_string(freq) + " > /sys/devices/system/cpu/cpufreq/policy2/scaling_max_freq";
  system(Command.c_str());
}

void run_graph(int pp1, int pp2, int big_freq, int little_freq) {
  scale_little(little_freq);
  scale_big(big_freq);

  int N_Frames = 20;

  char Run_Command[150];
  sprintf(Run_Command,
          "./graph_alexnet_all_pipe_sync --threads=4 --threads2=2 --target=NEON --n=%d --partition_point=%d "
          "--partition_point2=%d --order=B-G-L &> output.txt",
          N_Frames,
          pp1,
          pp2);
  system(Run_Command);
}

#define PARSEAMOUNT 2

array<string, PARSEAMOUNT> lineCharecteristics =
{
  "Frame rate is:",
  "Frame latency is:",
};

array<string, PARSEAMOUNT> outputLines =
{
  "fps:",
  "latency:",
};

string getOutput(int index, string text) {
  float floatOutput;
  if (stringstream(text) >> floatOutput) {
    return to_string(floatOutput);
  }
  return "";
}

// return 1 if target performance is met, 0 otherwise and write current to pointers
int check_performance(int target_fps, int target_latency, int* current_fps, int* current_latency) {
  ifstream myfile("/data/local/Working_dir/output.txt");

  int fps = 0;
  int latency = 0;

  /* Read Output.txt File and Extract Data */
  for (std::string line; getline(myfile, line);) {
    string temp;
    for (int i = 0; i < PARSEAMOUNT; i++) {
      if (line.find(lineCharecteristics[i]) == 0) {
        std::istringstream ss(line);
        while (!ss.eof()) {
          /* Extracting word by word from stream */
          ss >> temp;
          /* Checking the given word is float or not */
          string output = getOutput(i, temp);
          if (output != "") {
            if (i == 0) {
              fps = stoi(output);
            } else {
              latency = stoi(output);
            }
            break;
          }
        }
      }
    }
  }

  if (fps >= target_fps && latency <= target_latency) {
    return 1;
  } else {
    *current_fps = fps;
    *current_latency = latency;
    return 0;
  }
}

// optionally implement time
void govern(int target_latency, int target_fps, int population_size, int staleness_limit) {
  int adjusted_fps     = target_fps;
  int adjusted_latency = target_latency;

  int  stale_count     = 0;

  int pp1;
  int pp2;
  int big_freq;
  int little_freq;

  chromosome last_attempt;

  while (stale_count < staleness_limit) {
    // pass requirements to GA
    chromosome solution = genetic_algorithm(
          population_size,
          adjusted_latency,
          adjusted_fps,
          staleness_limit
        );

    // if solution is same as last solution, update staleness counter
    if (solution.fitness == last_attempt.fitness) {
      stale_count++;
    } else {
      last_attempt = solution;
      stale_count  = 0;
    }

    // get info from solution
    pp1         = solution.genes[0]->layers;
    pp2         = pp1 + solution.genes[1]->layers;
    big_freq    = BigFrequencyTable[solution.genes[0]->frequency_level];
    little_freq = LittleFrequencyTable[solution.genes[2]->frequency_level];

    printf("trying configuration: %d %d %d %d\n", pp1, pp2, big_freq, little_freq);
    run_graph(pp1, pp2, big_freq, little_freq);

    int current_fps;
    int current_latency;

    if (check_performance(target_fps, target_latency, &current_fps, &current_latency)) {
      cout << "Solution found." << endl;
      return;
    }

    string nudged = "";
    if (current_fps < target_fps) {
      adjusted_fps += (target_fps = current_fps) * NUDGE;
      nudged = "fps";
    } else if (current_latency > target_latency) {
      adjusted_latency -= (current_latency - target_latency) * NUDGE;
        nudged = "latency";
    }

    if (current_fps > target_fps * (1 + ANTI_NUDGE_THRESH)) {
      adjusted_fps -= (current_fps - target_fps*(1 + ANTI_NUDGE_THRESH)) * ANTI_NUDGE;
    } else if (current_latency < target_latency * (1 - ANTI_NUDGE_THRESH)) {
      adjusted_latency += (target_latency * (1 - ANTI_NUDGE_THRESH) - current_latency) * ANTI_NUDGE;
    }
    
    cout << "Configuration failed to reach " << nudged << " target\n";
  }

  cout << "Staleness limit reached\n";
}

int main(int argc, char** argv) {
  // Display help message
  cout << "Usage: " << argv[0] << " <Population_Size> <Target_Latency> <Target_FPS> <Staleness_Limit>\n";

  if (argc != 5) {
    std::cerr << "Error: Insufficient arguments. Please provide values for all parameters.\n";
    return EXIT_FAILURE;
  }

  // Parse population size
  int population_size = std::stoi(argv[1]);

  // Parse target latency
  int target_latency = std::stoi(argv[2]);

  // Parse target FPS
  int target_fps = std::stoi(argv[3]);

  // Parse staleness limit
  int staleness_limit = std::stoi(argv[4]);

  setup();

  govern(target_latency, target_fps, population_size, staleness_limit);
}

File: governor_model/governor.py
from GA import genetic_algorithm, chromosome_to_config
import subprocess
import time
import sys
from measurementAggregator import parseLine, transpose

ORDER = "B-G-L"
NUDGE = 1.3
REPETITION_LIMIT = 4

def govern(target_latency: float, target_fps: float):
    # Make sure testGovernor exists on the device
    adb_command = "adb shell"

    # Open a subprocess to communicate with ADB shell
    # with open("adb_pipe", "w") as pipe:
    process = subprocess.Popen(adb_command, shell=True, stdout=subprocess.PIPE, stderr=sys.stderr, stdin=subprocess.PIPE, text=True)

    # setup
    process.stdin.write("cd /data/local/Working_dir\n")
    process.stdin.write("export LD_LIBRARY_PATH=/data/local/Working_dir\n")
    process.stdin.write("echo performance > /sys/devices/system/cpu/cpufreq/policy0/scaling_governor\n")
    process.stdin.write("echo performance > /sys/devices/system/cpu/cpufreq/policy2/scaling_governor\n")
    process.stdin.write("echo 1 > /sys/class/fan/enable\n")
    process.stdin.write("echo 0 > /sys/class/fan/mode\n")
    process.stdin.write("echo 4 > /sys/class/fan/level\n")
    process.stdin.flush()

    adjusted_latency = target_latency
    adjusted_fps = target_fps
    win = False
    warm = None
    open("warmstart.txt", "w").close()
    # print("hi")
    last_attempt = ''
    stale_count = 0

    while not win:
        c = genetic_algorithm(100, adjusted_latency, adjusted_fps, 30, 40, save="force", warm=warm, save_location="warmstart.txt")
        if str(c) == last_attempt:
            stale_count += 1
        last_attempt = str(c)
        pp1, pp2, bfreq, lfreq = chromosome_to_config(c)
        print(f"\nTrying configuration:\npp1:{pp1}, pp2:{pp2}, Big frequency:{bfreq}, Small frequency:{lfreq}\n")
        process.stdin.write(f"echo {lfreq} > /sys/devices/system/cpu/cpufreq/policy0/scaling_max_freq\n") # little
        process.stdin.write(f"echo {bfreq} > /sys/devices/system/cpu/cpufreq/policy2/scaling_max_freq\n") # big
        # process.stdin.write("echo hi\n")
        # process.stdin.flush()
        # print(process.stdout.readline().strip())
        process.stdin.write(f"./graph_alexnet_all_pipe_sync --threads=4  --threads2=2 --n=20 --total_cores=6 --partition_point={pp1} --partition_point2={pp2} --order={ORDER} &> output.txt\n")
        process.stdin.write(f"./parse_perf\n")
        process.stdin.flush()

        try:
            while True:
                # Read the output from the ADB shell
                output = process.stdout.readline().strip()
                print("output is:", output)

                # Check if the output is not empty
                if output:
                    # Get the current timestamp at .001 second precision
                    timestamp = time.time()

                    result = f"[{timestamp}] {output}\n"
                    _, result = parseLine(result)
                    result = dict(transpose(result))
                    current_fps = result["fps"]
                    current_latency = result["latency"]
                    if current_fps >= target_fps and current_latency <= target_latency:
                        print("Solution found.")
                        win = True
                        process.terminate()
                        return
                    nudged = ""
                    if current_fps < target_fps:
                        adjusted_fps += (target_fps - current_fps) * NUDGE
                        nudged = "FPS"
                    if current_latency > target_latency:
                        adjusted_latency -= (current_latency - target_latency) * NUDGE
                        nudged = "latency"
                    warm = "warmstart.txt"

                    print(f"Configuration failed to reach {nudged} target.\n")
                    break

        except KeyboardInterrupt:
            # Handle keyboard interrupt (Ctrl+C) to stop the script
            print("Script terminated by user.")
            process.terminate()
            break

        if stale_count >= REPETITION_LIMIT:
            print("Governor can't find a better configuration.")
            process.terminate()
            break

    print("bye!")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("requires target latency and fps", file=sys.stderr)
        exit(0)
    try:
        target_latency = float(sys.argv[1])
        target_fps = float(sys.argv[2])
    except ValueError:
        print("targets must be numbers",  file=sys.stderr)
    #          lat          fps
    govern(target_latency, target_fps)
